docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -225,6 +225,205 @@ a :ref:`connector-page-sink-provider`.
+            if (isComplexExpressionPushdown(session)) {
+                List<String> newExpressions = new ArrayList<>();
+                List<ConnectorExpression> remainingExpressions = new ArrayList<>();
+                // Each expression can be broken down into a list of conjuncts
+                // joined with AND. We handle each conjunct separately.
+                for (ConnectorExpression expression : extractConjuncts(constraint.getExpression())) {
+                    // Try to convert the conjunct into something which is
+                    // understood by the underlying JDBC data source
+                    Optional<String> converted = jdbcClient.convertPredicate(session, expression, constraint.getAssignments());
+                    if (converted.isPresent()) {
+                        newExpressions.add(converted.get());
+                    }
+                    else {
+                        remainingExpressions.add(expression);
+                    }
+                }
+                // Calculate which parts of the expresion can be pushed down
+                // and which need to be calculated in Trino engine
+                newConstraintExpressions = ImmutableSet.<String>builder()
+                        .addAll(handle.getConstraintExpressions())
+                        .addAll(newExpressions)
+                        .build().asList();
+                remainingExpression = Optional.of(and(remainingExpressions));
+            }
+            else {
+                newConstraintExpressions = ImmutableList.of();
+                remainingExpression = Optional.empty();
+            }
+        }
+
+        // Return empty Optional if nothing changed in filtering
+        if (oldDomain.equals(newDomain) &&
+                handle.getConstraintExpressions().equals(newConstraintExpressions)) {
+            return Optional.empty();
+        }
+
+        handle = new JdbcTableHandle(
+                handle.getRelationHandle(),
+                newDomain,
+                newConstraintExpressions,
+                ...);
+
+        return Optional.of(
+                remainingExpression.isPresent()
+                        ? new ConstraintApplicationResult<>(handle, remainingFilter, remainingExpression.get(), precalculateStatisticsForPushdown)
+                        : new ConstraintApplicationResult<>(handle, remainingFilter, precalculateStatisticsForPushdown));
+    }
+
+In this example we are implementing a base class for many JDBC connectors,
+so we have to handle all the nuances of different JDBC-capable databases.
+For example, string comparison in some databases is case-insensitive, which
+means for those databases we can't push it down since it would change it's
+meaning.
+
+To handle push down for all the different types we need to ask our JDBC client
+for a ``ColumnMapping`` which can produce a ``PredicatePushdownController``
+which decides if a column domain can be pushed down.
```

Comment:
This is confusing. I'm not sure how I'd use this information when writing a non-JDBC connector. Note that I think we should have a whole separate page about JDBC connectors.